var/spack/repos/builtin/packages/xgboost/package.py

```diff
@@ -0,0 +1,97 @@
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+
+from spack import *
+import os
+from shutil import copyfile
+import glob
+
+
+class Xgboost(Package):
+    """
+    XGBoost is an optimized distributed gradient boosting library designed to
+    be highly efficient, flexible and portable. It implements machine learning
+    algorithms under the Gradient Boosting framework. XGBoost provides a
+    parallel tree boosting (also known as GBDT, GBM) that solve many data
+    science problems in a fast and accurate way. The same code runs on major
+    distributed environment (Hadoop, SGE, MPI) and can solve problems beyond
+    billions of examples.
+    """
+
+    homepage = "http://xgboost.readthedocs.io/en/latest/"
+    url      = "https://github.com/dmlc/xgboost"
+
+    version('0.7', commit='4aa346c', git="https://github.com/dmlc/xgboost", submodules=True)
+
+    variant('jvm-packages', default=False,
+            description='jvm-packages are compiled')
+
+    variant('cuda', default=False,
+            description='compiled with cuda support')
+
+    # cuda requires cmake v3.2+
+    depends_on('cmake@3.2:', type='build')
+    depends_on('maven', type='build', when='+jvm-packages')
+    depends_on('jdk', type='build', when='+jvm-packages')
+    depends_on('cuda', when='+cuda')
+
+    conflicts('%gcc@:4.7.4')
+    conflicts('%gcc@6.4:', when='+cuda')
+
+    def install(self, spec, prefix):
+        if '+cuda' in spec:
+            cmake('-DUSE_CUDA=ON')
+            # get back to xgboost dir to make
+            os.chdir(str(self.stage.source_path))
+        make()
+
+        mkdir(prefix.bin)
+        install('xgboost', prefix.bin)
+        install_tree('lib', prefix.lib)
+
+        # make jvm-packages
+        if '+jvm-packages' in spec:
+            with working_dir('jvm-packages'):
+                # custom repo location.
+                # Default is ~/.m2, and directory structure goes
+                # like ~/.m2/repository/ml/dmlc/...
+                # as opposed to m2/ml/dmlc/...
+                mvn_repo = str(self.stage.source_path) + '/jvm-packages/m2'
+                drepo = '-Dmaven.repo.local=' + mvn_repo
+                # compile with maven
+                mvn = which('mvn')
+                if self.run_tests:
+                    mvn(drepo, 'install')
+                else:
+                    mvn(drepo, 'install', '-DskipTests')
+
+            # save xgboost jars under package prefix
+            ver = str(self.spec.version)
+            for xgtype in ['', '-spark', '-flink', '-example']:
+                ujars = glob.glob(mvn_repo + '/ml/dmlc/xgboost4j' + xgtype +
+                                  '/' + ver + '/*.jar')
+                for jar in ujars:
+                    copyfile(jar, str(spec.prefix) + '/' +
```

Comment:
I would use the Spack function `install()` here. It wraps around `shutil.copy` and also handles some file permission settings. You should also check out `join_path()`.